sw/device/lib/base/memory.c

```diff
@@ -60,14 +170,40 @@ int OT_PREFIX_IF_NOT_RV32(memcmp)(const void *lhs, const void *rhs,
 OT_WEAK
 int OT_PREFIX_IF_NOT_RV32(memrcmp)(const void *lhs, const void *rhs,
                                    size_t len) {
-  const uint8_t *lhs8 = (uint8_t *)lhs;
-  const uint8_t *rhs8 = (uint8_t *)rhs;
-  size_t j;
-  for (size_t i = 0; i < len; ++i) {
-    j = len - 1 - i;
-    if (lhs8[j] < rhs8[j]) {
+  if (len == 0) {
+    return kMemCmpEq;
+  }
+  const unsigned char *lhs8 = (const unsigned char *)lhs;
+  const unsigned char *rhs8 = (const unsigned char *)rhs;
+  size_t leading_bytes_end, aligned_words_end;
+  compute_alignment((uintptr_t)lhs, (uintptr_t)rhs, len, &leading_bytes_end,
+                    &aligned_words_end);
+  size_t end = len;
+  for (; end > aligned_words_end; --end) {
+    const size_t index = end - 1;
+    if (lhs8[index] < rhs8[index]) {
+      return kMemCmpLt;
+    } else if (lhs8[index] > rhs8[index]) {
+      return kMemCmpGt;
+    }
+  }
+  for (; end > leading_bytes_end; end -= sizeof(uint32_t)) {
+    const size_t index = end - sizeof(uint32_t);
```

Comment:
If you agree with the comment above, we should decrement before the remaining 2 loops, e.g. `for(i -= sizeof(uint32_t); i >= leading_bytes_end; i -= sizeof(uint32_t)) {`